Bluetooth: L2CAP: Allow finer control over receiving credits

Vudentz opened the pull request:
These changes add the possibility to control rx credits, it adds a return to recv which can be set to -EINPROGRESS to process packets asynchronously and then confirm with reception with bt_l2cap_chan_recv. In addition to this it also removes the use of a threshold to updates the rx credits and replaces with a simpler logic of restoring the init_credits after each packet is reassembled.

Vudentz committed:
Bluetooth: L2CAP: Add return to recv

This adds a int return to recv callback which can be used to notify the
stack about errors when receiving a packet. In addition to that the user
can return -EINPROGRESS to inform the stack the data will be processed
asynchronously which can be complete by calling
bt_l2cap_chan_recv_complete.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: L2CAP: Restore credits after receiving an SDU

This simplify the logic of restoring the credits after each SDU instead
of using an arbitrary threshold which was not configurable per channel.

Because the credits are restored only when the full SDU has been
reassembled it means the channels needs to be set up with enough for
the configured MTU otherwise there is a risk of the channel to run out
of RX credits before the packet is fully reassembled, because of such
corner case the code will now warn if a channel is setup with not enough
init_credits.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: L2CAP: Fix identation of struct members

Align identation of struct members.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

thomas-ot commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -35,7 +35,8 @@
 #define L2CAP_LE_MAX_CREDITS		(CONFIG_BT_RX_BUF_COUNT - 1)
 #endif
 
-#define L2CAP_LE_CREDITS_THRESHOLD(_creds) (_creds / 2)
```
With _creds / N, you can only set the threshold to 100% (send credit on every packet received), 50%, 33%, 25%, ...
If I would change the threshold from the default 50%, I would choose something higher to improve bandwidth, but not 100%, since this would generate too much credit traffic.
Some platforms don’t support efficient division.

For these reasons, I suggest that 

```
#define L2CAP_LE_CREDITS_THRESHOLD(_creds) \
	(_creds / CONFIG_BT_L2CAP_RX_CREDIT_THRESHOLD)
```
is changed to:

```
#define L2CAP_LE_CREDITS_THRESHOLD(_creds) \
	(_creds - (_creds >> CONFIG_BT_L2CAP_RX_CREDIT_THRESHOLD)
```

I.e. you subtract 2^-(N) of 100%, meaning that for N=0,1,2,3, you get a threshold of 100%, 50%, 75%, 87.5%.

Since most platforms do mult efficiently, it is probably better to do this:

```
#define L2CAP_LE_CREDITS_THRESHOLD(_creds) \
	((_creds * CONFIG_BT_L2CAP_RX_CREDIT_THRESHOLD) >> 8)
```

This would give e.g. these thresholds:
0	0.0%
64	25.0%
128	50.0%
192	75.0%
256	100.0%

Maybe it would be simpler and more flexible to have a callback to get the threshold from the channel? By having a CONFIG, you decide for all channels globally, which may not be what you want. There are also many configs already :-)

Vudentz replied:
Yep, Ive been debating to add a callback to read the credit threshold, perhaps with that we don't need to add set_credits function and just leave it up for the channel owner. I will play with this idea and let you know.

thomas-ot commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1340,6 +1341,10 @@ static void l2cap_chan_update_credits(struct bt_l2cap_le_chan *chan,
 
 	/* Restore credits */
 	credits = chan->rx.init_credits - k_sem_count_get(&chan->rx.credits);
```
If the channel sets init_credits to 0, then credits could become negative, but credits is unsigned.
Or how did you intend the channel to use bt_l2cap_chan_set_credits()?

Vudentz replied:
Yep, the it should never reach this line since the check above it should return iirc.

thomas-ot commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1355,6 +1360,17 @@ static void l2cap_chan_update_credits(struct bt_l2cap_le_chan *chan,
 	BT_DBG("chan %p credits %u", chan, k_sem_count_get(&chan->rx.credits));
 }
 
+u16_t bt_l2cap_chan_set_credits(struct bt_l2cap_chan *chan, u16_t credits)
```
It is confusing to me that the param is called credits when what the function does is to set init_credits, i.e. you are not giving credits in a relative way (like LE Flow Control Credit packet). As hinted above, I'm unsure of how to use this function in practice.

Vudentz replied:
It is used to restore the credits, though perhaps it clearer if we do say init_credits instead, we cannot take back credits so that can only affect the next time we update the credits so you call it whenever you want to change the init_credits while connected e.g. if you run out of buffer set it to 0 which will likely make the remote stop to transmit, then as the buffers return to the pool you may restore the credits.

thomas-ot replied:
The credits are counted in L2CAP MPS (PDU), but the alloc_buf callback is supposed to allocate an SDU (L2CAP MTU). So, how will I now how many credits to give after I set init_credits=0?
Further, can I be sure not to get a recv() callback after setting init_credits=0, i.e. could there be packets in the pipeline?

The way it is done in NimBLE seems like a good design, where the credit is hidden from the application. The application just calls 
````
ble_l2cap_recv_ready(*chan, sdu_rx) 
````
when it is able to receive another SDU. Until then, the stack buffers as much as its pool allows it.

It may still be a good idea to keep the set_init_credits in order to dynamically re-prioritize the buffer usage to different channels.

 https://github.com/apache/mynewt-nimble/blob/1f7227ed860c8783ec874effde4c0bd620977187/nimble/host/src/ble_l2cap_coc.c#L449
https://www.mail-archive.com/dev@mynewt.incubator.apache.org/msg02418.html

Vudentz committed:
Bluetooth: L2CAP: Prevent the channel to stall

If the remote stack is not able to fully utilize each segment it is
possible it would run out of credits before completing the SDU, these
changes detects if that would happen and attempt to restore enough
credits for the SDU to be received.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

thomas-ot commented:
The latest changes look good to me.

The changes seem to make it possible to get flow control between the application and the stack without blocking the RX thread of the BT host. Flow control is required when the application can't keep up with the incoming packets, e..g. if the processing of the packet content takes a long time and/or the data needs to be transferred over serial to another processor.

I'm looking forward to reading the documentation on the intended use of the API (as discussed offline).

I'm a little in doubt if a credit packet is sent each time the application has finished processing a packet and calls bt_l2cap_chan_recv(), after first having returned EINPROGRESS in the recv callback. Does this lead to a credit packet for each received packet? If so, this is probably not the optimal solution for most applications.

Vudentz committed:
Bluetooth: L2CAP: Make use of net_buf_append_bytes

This uses net_buf_append_bytes to reassemble the SDU segments instead of
doing it manually.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
net: buf: Use net_buf_add_mem in net_buf_append_bytes

This ensures the memory copies are properly logged when debug is
enabled.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@thomas-ot it won't return more credits until the application calls bt_l2cap_chan_recv, note that you may still received more SDU in the meantime if there are still credits left and the packets don't need more that what is left, in that case Id assume the application will keep return -EINPROGRESS until it is done processing them.

thomas-ot commented:
> @thomas-ot it won't return more credits until the application calls bt_l2cap_chan_recv, note that you may still received more SDU in the meantime if there are still credits left and the packets don't need more that what is left, in that case Id assume the application will keep return -EINPROGRESS until it is done processing them.

I understand and agree that the app should keep returning EINPROGRESS if it cannot keep up. 
Let's say the app does this for 3 SDUs without calling bt_l2cap_chan_recv. Finally, it finishes 1 out of the 3 and calls bt_..._recv. This triggers the stack to send a credit message to the peer, right? This may or may not be optimal, i.e. maybe it's better (for bandwidth util) to wait until the 2nd or 3rd SDU queued up has been processed. It would be nice if the interface to tell the stack "I have processed the SDU you gave me" could be split from the policy of when to send credits to the peer.

Vudentz commented:
> I understand and agree that the app should keep returning EINPROGRESS if it cannot keep up.
Let's say the app does this for 3 SDUs without calling bt_l2cap_chan_recv. Finally, it finishes 1 out of the 3 and calls bt_..._recv. This triggers the stack to send a credit message to the peer, right? This may or may not be optimal, i.e. maybe it's better (for bandwidth util) to wait until the 2nd or 3rd SDU queued up has been processed. It would be nice if the interface to tell the stack "I have processed the SDU you gave me" could be split from the policy of when to send credits to the peer.

Perhaps you are suggesting that we calculate the amount of credits to return based on each packet, that way it return just the exact number of credits processed.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1351,12 +1347,62 @@ static void l2cap_chan_update_credits(struct bt_l2cap_le_chan *chan,
+	BT_DBG("chan %p buf %p", chan, buf);
+
+	/* Restore credits used by packet */
+	credits = (net_buf_frags_len(buf) + (ch->rx.mps + 1)) / ch->rx.mps;
```
Is it +1 because you simplified this?
net_buf_frags_len(buf) + 2 + (ch->rx.mps - 1)) / ch->rx_mps
If so I think it is better to write it like that.

Vudentz replied:
Yep it is counting the SDU length I will add that.

joerchan commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1386,36 +1430,42 @@ static void l2cap_chan_le_recv_sdu(struct bt_l2cap_le_chan *chan,
+	chan->_sdu_len = 0;
+
+	if (err < 0) {
+		if (err != -EINPROGRESS) {
```
Should you not do this for line 1513 as well?

Vudentz replied:
Nope, that is not doing segmentation in that case the channel will receive each segment and has to reassemble it.

Vudentz committed:
Bluetooth: shell: Add l2cap-recv command

This adds l2cap-recv which can be used to set a delay when confirming
the receptions of packets:

> l2cap-recv 1000
Incoming conn 0x00401c20
Channel 0x004220c8 connected
Channel 0x004220c8 requires buffer
Incoming data channel 0x004220c8 len 230
00000000 00 00 00 00 e6 00 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
00000010 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
00000020 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
00000030 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
00000040 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
00000050 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
00000060 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
00000070 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
00000080 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
00000090 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
000000a0 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
000000b0 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
000000c0 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
000000d0 7f 7f 7f 7f 7f 7f 7f 7f  7f 7f 7f 7f 7f 7f 7f 7f
000000e0 7f 7f 7f 7f 7f 7f
Delaying response...
Confirming reception

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@joerchan This should address all the problems you had commented, can you have a look.

joerchan commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -1833,6 +1871,15 @@ static int cmd_l2cap_send(int argc, char *argv[])
+{
+	if (argc > 1) {
+		l2cap_recv_delay = strtoul(argv[1], NULL, 10);
+	}
```
else recv_delay = 0

Vudentz replied:
I prefer to have it print the current delay if no parameter is given.

thomas-ot commented:
> Perhaps you are suggesting that we calculate the amount of credits to return based on each packet, that way it return just the exact number of credits processed.

It was not what I meant, but that is good thing.

The small issue I see is that the sending of credit packets is tied to the application giving credit back. Before this PR, there was a policy to send a credit packet when the credits were below 50% of init_credits, i.e. the stack would not necessarily send a credit packet to the peer on each received SDU. E.g. if init_credits were 10, it could receive 5 packets before a credit packet would be sent to the peer. Now, it seems a credit packet will be sent every time the application calls bt_l2cap_chan_recv_complete(), thus wasting bandwidth compared with the previous behavior.
I don't think this issue should hold the PR back, since the critical missing feature is having the flow control between the stack and the application.
